_posts/2018-03-15-complete-introduction-to-nodejs.md

```diff
@@ -0,0 +1,406 @@
+    } else {
+	// if no data was passed to the callback, we'll simply log 'Hello world'
+	console.log('Hello world');
+    }
+});
+
+// now that we are listening for the 'hello' event, we'll emit the event, once with data, and once without data.
+eventEmitter.emit('hello', {name: 'Mike'}); // logs 'Hello Mike'
+eventEmitter.emit('hello'); // logs 'Hello world'
+```
+
+## Streams
+There's many ways that you can utilise readable/writable streams with NodeJs, for example the file system to read/write to files.
+But to give you a simple example, let's reuse our code from our hello server.
+Since the request object is a readable stream and the response object is a write-able stream,
+we can create an application that pipes the data from the request, back to the response.
+Let's see it in action with an example:
+
+`streams.js`
+
+```javascript
+const http = require('http');
+
+const server = http.createServer((request, response)=> {
+  res.writeHead(200, {'Content-Type': 'text/plain'});
+  // the request is a readable stream, this means that the connection isn't immediately closed,
+  // the connection stays open for as long as the client keeps sending data.
+  // Streams inherit from the event emitter, so we can listen for the request stream's 'readable' and 'end' events
+  // the readable event is triggered whenever the request has sent a chunk of data that can be read.
+  req.on('readable', ()=>{
+      let chunk = null;
+      // as long as we can read chunks from the request, we write those chunks to the response.
+      while(null !== (chunk = request.read())){
+         // we can keep writing to the writable response stream as long as the connection is open,
+         // so we keep piping the readable data to the response
+          response.write(chunk);
+      }
+      // you can test it with curl from the terminal by sending a 'hello' string as data,
+      // simply run: curl -d 'hello' http://localhost:3000
+      // the 'hello' string is being sent back to the client
+  });
+
+  req.on('end', ()=>{
+      // once the request stream is closed, we also close the response stream
+      response.end();
+  });
+
+  /*
+	I've written it out completely to show you what is does,
+	but in fact we can use the pipe method to refactor the request's 'on readable' in some much simpler code
+	req.on('readable', ()=>{
+             request.pipe(response)
+        })
+   */
+});
+
+server.listen(3000, ()=>{
+    console.log('Server running at localhost:3000/');
+});
+```
+
+## Cluster
+A single instance of Node.js runs in a single thread.
+To take advantage of multi-core systems, the user will sometimes want to launch a cluster of Node.js processes to handle the load.
+
+```javascript
+// require the cluster module
+const cluster = require('cluster');
+// we'll set up a http server on all cpu's and load balance between them
+const http = require('http');
+// we need to know the amount of cpu's our machine has available,
```

Comment:
cpus

Reply:
fixed in https://github.com/ordina-jworks/ordina-jworks.github.io/pull/359/commits/daf5636773b89b76532c425468fddc6511bf0643